### PR TITLE
WT-6914 Write database files subpage for architecture guide

### PR DESCRIPTION
**Pull request by mm-ng**

### CODE REVIEW

**Review by ddanderson — COMMENTED** [review]
@mm-ng, this is a great start.  I think you've covered this in about the right depth.  I've made a few specific suggestions.  I was thinking that it would be good to mention a few of the important functions used in the creation/reading of data files:

Low level functions that pack or unpack the contents of individual cells are in `cell_inline.h` .  Functions like `__wt_cell_unpack_kv` and `__wt_cell_unpack_addr` to unpack keys values and addresses are found here, and uses of such functions appear in the BTree code.  Similarly, there are `__wt_cell_pack_*` functions to pack particular kinds of keys, values and addresses; these are used in the reconciliation code.

You can use my wording (or not).  This change basically LGTM, but I'll take another review pass when you're ready.

File: src/docs/arch-data-file.dox [hunk 1]
```diff
@@ -1,17 +1,90 @@
...
+is made up of data cells (WT_CELL_VALUE/VALUE_COPY/VALUE_OVFL) or 
+deleted cells (WT_CELL_DEL)
+
+(3) Overflow Pages - overflow pages are needed when keys/values that are 
+too large must be stored separately in the file, apart from where the item 
+is logically placed. Page sizes and configuration values such as 
+internal_key_max, leaf_key_max and leaf_value_max are used to determine 
+overflow items.
+
+Whilst the above gives a general overview of how different page types are
```
Comment (ddanderson) [inline src/docs/arch-data-file.dox:84]: Whilst -> while  (whilst is archaic in US English).  I think this appears in a few spots.

File: src/docs/arch-data-file.dox [hunk 2]
```diff
@@ -1,17 +1,90 @@
...
-can represent multiple versions.  Extremely large values may be represented
-as a reference to another page.
-
-The exact encoding is rather complex, and beyond what can be described here.
-The encoding strikes a balance between data that can be compacted efficiently
-in time and space, extensibility, and compatibility with previous versions.
+Database Files
+
+WiredTiger stores databases in a durable format through writing data to the 
+physical disk/s. As part of this process, WiredTiger creates a number of
```
[the remark this answers is not in the document]
Comment (mm-ng) [reply src/docs/arch-data-file.dox:6]: Yes that makes sense, I will change this.

File: src/docs/arch-data-file.dox (hunk 1 above)
Comment (mm-ng) [reply src/docs/arch-data-file.dox:84]: Thanks for the heads up, I didn't know this!

File: src/docs/arch-data-file.dox [hunk 3]
```diff
@@ -1,17 +1,90 @@
...
+is logically placed. Page sizes and configuration values such as 
+internal_key_max, leaf_key_max and leaf_value_max are used to determine 
+overflow items.
+
+Whilst the above gives a general overview of how different page types are 
+structured, it is a simplified representation. Due to the large amount of 
+data being stored, WiredTiger may compress data to preserve memory and disk 
+space. Keys may not be stored in their entirety when prefix compression 
+is used (the identical key prefix is stored only once per page to reduce 
+size).
```
[the remark this answers is not in the document]
Comment (mm-ng) [reply src/docs/arch-data-file.dox:89]: Good point.

**Comment by mm-ng** [comment]
@ddanderson Thanks so much for all the feedback. I have incorporated your suggestions and also added a section on the data file functions. Let me know if there is anything else you think would be worth discussing in the article.

File: src/docs/arch-data-file.dox [hunk 4]
```diff
@@ -1,17 +1,105 @@
...
+
+Different types of pages are made up of different types of cells.
+
+(1) Internal Pages - as the B-Tree grows in size and layer/s are added,
+there will be pages between the root page and leaf page. These internal
+pages contain keys that point to other pages.
+- The row-store internal page \c WT_PAGE_ROW_INT is made up of a key cell
+(\c WT_CELL_KEY or \c WT_CELL_KEY_OVFL) followed by an offpage-reference
+(\c WT_CELL_ADDR_XXX).
+- The column-store internal page simply contains an off-page reference.
```
Comment (ddanderson) [inline src/docs/arch-data-file.dox:69]: Is this true?  References don't have a corresponding starting (recordid) key?

File: src/docs/arch-data-file.dox [hunk 5]
```diff
@@ -1,17 +1,105 @@
...
+
+(2) Leaf Pages - leaf pages consist of a page header, followed by keys,
+values or addresses (these reference overflow pages).
+- The row-store leaf page \c WT_PAGE_ROW_LEAF is made up of a key cell
+(\c WT_CELL_KEY or \c WT_CELL_KEY_OVFL) followed by a value
+(\c WT_CELL_VALUE/VALUE_COPY/VALUE_OVFL).
+- The column-store leaf page (for variable-length data) WT_PAGE_COL_VAR
+is made up of data cells (WT_CELL_VALUE/VALUE_COPY/VALUE_OVFL) or
+deleted cells (WT_CELL_DEL)
```
Comment (ddanderson) [inline src/docs/arch-data-file.dox:78]: If there is no record_id key, how is it deduced?  (There must be a starting id somewhere).

**Review by ddanderson — APPROVED** [review]
LGTM.  I have some remaining questions about how column keys are deduced.

File: src/docs/arch-data-file.dox (hunk 4 above)
Comment (mm-ng) [reply src/docs/arch-data-file.dox:69]: What I meant by this sentence is I thought the column-store internal page only had a WT_CELL_ADDR_<> cell. I believe the recordid is part of the WT_CELL_COMMON_FIELDS which is shared across all cell types. Would it be helpful to talk about the structure of cells?

Comment (ddanderson) [reply src/docs/arch-data-file.dox:69]: You're right, it looks like internal pages just have references (presumably the key needed for searching can be deduced by looking at the first key on each page referenced).

File: src/docs/arch-data-file.dox (hunk 5 above)
Comment (ddanderson) [reply src/docs/arch-data-file.dox:78]: You might add: `The page header has a starting record number, so the associated keys for all column store values is deduced by their position in the page.`

**Comment by ddanderson** [comment]
@mm-ng, I've asked you to insert one sentence.  Everything else looks great, ready to merge.

File: src/docs/arch-data-file.dox (hunk 5 above)
Comment (mm-ng) [reply src/docs/arch-data-file.dox:78]: Thanks for this suggestion Don. I have added it in and will make the merge.

**Pull request merged**